src/MagnumPlugins/BasisImporter/BasisImporter.cpp

```diff
@@ -0,0 +1,259 @@
+        }
+    }
+}
+
+BasisImporter::~BasisImporter() = default;
+
+auto BasisImporter::doFeatures() const -> Features { return Feature::OpenData; }
+
+bool BasisImporter::doIsOpened() const { return _in; }
+
+void BasisImporter::doClose() { _in = nullptr; }
+
+void BasisImporter::doOpenData(const Containers::ArrayView<const char> data) {
+    /* Because here we're copying the data and using the _in to check if file
+       is opened, having them nullptr would mean openData() would fail without
+       any error message. It's not possible to do this check on the importer
+       side, because empty file is valid in some formats (OBJ or glTF). We also
+       can't do the full import here because then doImage2D() would need to
+       copy the imported data instead anyway (and the uncompressed size is much
+       larger). This way it'll also work nicely with a future openMemory(). */
+    if(data.empty()) {
+        Error{} << "Trade::BasisImporter::openData(): the file is empty";
+        return;
+    }
+
+    if(!_state->_transcoder.validate_header(data.data(), data.size())) {
+        Error() << "Trade::BasisImporter::openData(): invalid header";
+        return;
+    }
+
+    if(!_state->_transcoder.start_transcoding(data.data(), data.size())) {
+        Error() << "Trade::BasisImporter::openData(): bad basis file";
+        return;
+    }
+
+    _in = Containers::Array<unsigned char>(data.size());
+    std::copy(data.begin(), data.end(), _in.begin());
+}
+
+UnsignedInt BasisImporter::doImage2DCount() const {
+    return _state->_transcoder.get_total_images(_in.data(), _in.size());
+}
+
+Containers::Optional<ImageData2D> BasisImporter::doImage2D(UnsignedInt index) {
+    const UnsignedInt level = 0;
+
+    const std::string targetFormatStr = configuration().value<std::string>("format");
+    if(targetFormatStr.empty()) {
+        Error() << "Trade::BasisImporter::image2D(): no format to transcode "
+            "to was specified. Either load the plugin via one of its BasisImporterEtc1, ... "
+            "aliases, or set the format explicitly via plugin configuration.";
+        return {};
+    }
+
+    const BasisTranscodingType targetFormat =
+        configuration().value<BasisTranscodingType>("format");
+    if(Int(targetFormat) == -1) {
+        Error() << "Trade::BasisImporter::image2D(): invalid transcoding target format"
+            << targetFormatStr.c_str() << Debug::nospace << ", expected to be one of:"
+            << "Etc1, Etc2, Bc1, Bc3, Bc4, Bc5, Bc7M6OpaqueOnly, Pvrtc1_4OpaqueOnly";
+        return {};
+    }
+    const basist::transcoder_texture_format format =
+        TranscoderTextureFormat[Int(targetFormat)];
+    const UnsignedInt bytes_per_block = basis_get_bytes_per_block(format);
+
+    basist::basisu_image_info info;
+    if (!_state->_transcoder.get_image_info(_in.data(), _in.size(), info, index)) {
+        Error{} << "Trade::BasisImporter::image2D(): unable to get image info";
+    }
```

Comment:
Missing `return {};`? :scream: 

It's like the Murphy's law -- the only code path we don't test because we think "it's good enough" will have a bug in it :D Looking at the source of `get_image_info()` I have no idea how to trigger this one (most of the sanity checks done there are checked already by `AbstractImporter` base implementation itself), so probably just add the `return` and pray :)